flow/flow.go

```diff
@@ -0,0 +1,303 @@
+	// Abort specific flow execution by request id
+	Abort(flow string, reqID string) error
+	// State show state specific flow execution by request id
+	State(flow string, reqID string) (string, error)
+	// Result get result of the flow step
+	Result(flow string, reqID string, step *Step) ([]byte, error)
+	// Stop executor and drain active workers
+	Stop() error
+}
+
+type Step struct {
+	// name of step
+	ID string
+	// Retry count for step
+	Retry int
+	// Timeout for step
+	Timeout int
+	// Step operation to execute
+	Operation Operation
+	// Which step use as input
+	Input string
+	// Where to place output
+	Output string
+	// Steps that are required for this step
+	After []string
+	// Steps for which this step required
+	Before []string
+	// Step operation to execute in case of error
+	Fallback Operation
+}
+
+func (s *Step) Name() string {
+	return s.ID
+}
+
+func (s *Step) Id() string {
+	return s.ID
+}
+
+func (s *Step) String() string {
+	return s.ID
+	//return fmt.Sprintf("step %s, ops: %s, requires: %v, required: %v", s.ID, s.Operations, s.Requires, s.Required)
+}
+
+type Option func(*Options)
+
+type Options struct {
+	// Number of worker goroutines
+	Concurrency int
+	// Preallocate worker goroutines
+	Prealloc bool
+	// If no workers available no blocking
+	Nonblock bool
+	// Wait completiong before stop
+	Wait bool
+	// StateStore is used for flow state marking
+	StateStore DataStore
+	// DataStore is used for intermediate data passed between flow nodes
+	DataStore DataStore
+	// FlowStore is used for storing flows
+	FlowStore FlowStore
+	// EventHandler is used to notification about flow progress
+	EventHandler EventHandler
+	// PanicHandler is used for recovery panics
+	PanicHandler func(interface{})
+	// Logger is used internally to provide messages
+	Logger logger.Logger
+	// Context is used for storing non default options
+	Context context.Context
+}
+
+type flowKey struct{}
+
+func FlowToContext(ctx context.Context, fl Flow) context.Context {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	return context.WithValue(ctx, flowKey{}, fl)
+}
+
+func FlowFromContext(ctx context.Context) (Flow, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("invalid context")
+	}
+	flow, ok := ctx.Value(flowKey{}).(Flow)
+	if !ok {
+		return nil, fmt.Errorf("invalid context")
+	}
+	return flow, nil
+}
+
+type ExecuteOptions struct {
+	// Passed request id
+	ID string
+	// Passed step to start swafrom
+	Step string
+	// Output which step output returns
+	Output string
+	// Timeout for currenct execition
+	Timeout time.Duration
+	// Async execution run
+	Async bool
+	// Concurrency specify count of workers create for nodes in flow
+	Concurrency int
+	// Retries specify count of retries in case of node execution failed
+	Retries int
+	// Context is used for storing non default options
+	Context context.Context
+	// Client for communication
+	Client client.Client
+	// Broker for communication
+	Broker broker.Broker
```

Comment:
If you have the Client you don't need the broker because the Client contains the broker.

Reply:
removed from options, thanks